core/src/main/java/com/linecorp/armeria/common/RequestContext.java

```diff
@@ -435,6 +435,33 @@ default ByteBufAllocator alloc() {
     @MustBeClosed
     SafeCloseable push();
 
+    /**
+     * Adds a hook which is invoked whenever this {@link RequestContext} is pushed to the thread-local stack.
+     * The {@link SafeCloseable} returned by {@code onPush} will be called whenever this
+     * {@link RequestContext} is popped from the thread-local stack.
+     * This method is useful when you need to propagate a custom context in this {@link RequestContext}'s scope.
+     * For example:
+     * <pre>{@code
+     * ctx.hook(() -> {
+     *     MDC.put("transactionId", "1234567890");
+     *     // Should clean up the thread local resources when closing.
+     *     return () -> MDC.remove("transactionId");
+     * });
+     *
+     * <p>Note that this operation is highly performance-sensitive operation, and thus
+     * it's not a good idea to run a time-consuming task.
+     * }</pre>
+     */
+    void hook(Supplier<? extends SafeCloseable> contextHook);
```

Comment:
Instead of this method which we've talked about a lot before, maybe it's better to add an SPI of `RequestContextWrapper`? Can call `decorateScope(get(), Scope.NOOP)` I think even outside the code of the decoratos

Reply:
> maybe it's better to add an SPI of RequestContextWrapper?

Sorry, I didn't understand. Did you mean `RequestContextStorage` or introduce a new SPI for `RequestContextWrapper`?

Reply:
I guess he meant an SPI for hooks. We have an SPI for storage, but we don't have it for hooks. We could automatically install a hook when armeria-brave is loaded for example.